pkg/trace/agent/agent.go

```diff
@@ -484,11 +475,37 @@ func isManualUserDrop(priority sampler.SamplingPriority, pt *traceutil.Processed
 	return dm == manualSampling
 }
 
-// sample reports the number of events found in pt and whether the chunk should be kept as a trace.
-// sample does a semi-deep copy of pt to avoid making accidental changes to which spans are in the trace.
-// But any changes made directly to the spans, such as setting tags, etc. is not allowed.
-func (a *Agent) sample(now time.Time, ts *info.TagStats, pt *traceutil.ProcessedTrace) (numEvents int64, keep bool, retPt *traceutil.ProcessedTrace) {
-	pt = pt.Clone()
+// sample performs all sampling on the processedTrace modifying it as needed and returning if the trace should be kept and the number of events in the trace
+func (a *Agent) sample(now time.Time, ts *info.TagStats, pt *traceutil.ProcessedTrace) (keep bool, numEvents int) {
+	// We have a `keep` that is different from pt's `DroppedTrace` field as `DroppedTrace` will be sent to intake.
+	// For example: We want to maintain the overall trace level sampling decision for a trace with Analytics Events
+	// where a trace might be marked as DroppedTrace true, but we still sent analytics events in that ProcessedTrace.
+	keep, skipAnalyticsEvents := a.traceSampling(now, ts, pt)
+
+	var events []*pb.Span
+	if !skipAnalyticsEvents {
+		// We skip analytics events when a trace is marked as manual drop (aka priority -1)
+		// Note that we DON'T skip single span sampling. We only do this for historical
+		// reasons and analytics events are deprecated so hopefully this can all go away someday.
```

Comment:
I think ideally this comment should be in `traceSampling` where we actually decide the value of `skipAnalyticsEvents`. Doesn't hurt to duplicate the comment as well and keep it in both functions if you prefer!